### PR TITLE
raftstore: add cmd observer

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

###  What have you changed?

Add cmd observer to observe the executed commands of apply worker.

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.
###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

File: src/raftstore/store/fsm/apply.rs [hunk 1]
```diff
@@ -304,6 +305,8 @@ struct ApplyContext {
     sync_log_hint: bool,
     // Whether to use the delete range API instead of deleting one by one.
     use_delete_range: bool,
+
+    cmd_batches: Vec<CmdBatch>,
```
Comment (overvenus) [inline src/raftstore/store/fsm/apply.rs]: Could you move it into cmd observer?

File: src/raftstore/coprocessor/dispatcher.rs [hunk 2]
```diff
@@ -246,7 +252,7 @@ impl CoprocessorHost {
     }
 
     /// Call all pre apply hook until bypass is set to true.
-    pub fn pre_apply(&self, region: &Region, req: &RaftCmdRequest) {
+    pub fn pre_apply(&self, region: &Region, index: u64, req: &RaftCmdRequest) {
```
Comment (overvenus) [inline src/raftstore/coprocessor/dispatcher.rs]: In fact, we don't need to add the index.

File: src/raftstore/coprocessor/dispatcher.rs [hunk 3]
```diff
@@ -261,28 +267,32 @@ impl CoprocessorHost {
...
                 &self.registry.admin_observers,
                 pre_apply_admin,
+                index,
                 admin
             );
         }
     }
 
-    pub fn post_apply(&self, region: &Region, resp: &mut RaftCmdResponse) {
+    pub fn post_apply(&self, region: &Region, index: u64, resp: &mut RaftCmdResponse) {
```
Comment (overvenus) [inline src/raftstore/coprocessor/dispatcher.rs]: In fact, we don't need to add the index.

File: src/raftstore/coprocessor/dispatcher.rs [hunk 4]
```diff
@@ -424,7 +449,12 @@ mod tests {
             ctx.bypass = self.bypass.load(Ordering::SeqCst);
         }
 
-        fn post_apply_query(&self, ctx: &mut ObserverContext<'_>, _: &mut Vec<Response>) {
+        fn post_apply_query(
+            &self,
+            ctx: &mut ObserverContext<'_>,
+            _: &RaftResponseHeader,
```
Comment (overvenus) [inline src/raftstore/coprocessor/dispatcher.rs]: We don't need to add `RaftResponseHeader`.

File: src/raftstore/coprocessor/dispatcher.rs [hunk 5]
```diff
@@ -394,12 +413,18 @@ mod tests {
...
             self.called.fetch_add(2, Ordering::SeqCst);
             ctx.bypass = self.bypass.load(Ordering::SeqCst);
         }
 
-        fn post_apply_admin(&self, ctx: &mut ObserverContext<'_>, _: &mut AdminResponse) {
+        fn post_apply_admin(
+            &self,
+            ctx: &mut ObserverContext<'_>,
+            _: u64,
+            _: &RaftResponseHeader,
```
Comment (overvenus) [inline src/raftstore/coprocessor/dispatcher.rs]: We don't need to add `u64` and `RaftResponseHeader`.

File: src/raftstore/coprocessor/dispatcher.rs [hunk 6]
```diff
@@ -394,12 +413,18 @@ mod tests {
             Ok(())
         }
 
-        fn pre_apply_admin(&self, ctx: &mut ObserverContext<'_>, _: &AdminRequest) {
+        fn pre_apply_admin(&self, ctx: &mut ObserverContext<'_>, _: u64, _: &AdminRequest) {
```
Comment (overvenus) [inline src/raftstore/coprocessor/dispatcher.rs]: We don't need to add `u64`.

File: src/raftstore/store/fsm/apply.rs [hunk 7]
```diff
@@ -337,22 +340,41 @@ impl ApplyContext {
...
             self.kv_wb = Some(WriteBatch::with_capacity(DEFAULT_APPLY_WB_SIZE));
             self.kv_wb_last_bytes = 0;
             self.kv_wb_last_keys = 0;
         }
         self.cbs.push(ApplyCallback::new(delegate.region.clone()));
         self.last_applied_index = delegate.apply_state.get_applied_index();
+
+        if let Some(enabled) = &delegate.cmd_observer_enabled {
+            let region_id = delegate.region_id();
+            if enabled.load(Ordering::Relaxed) {
```
Comment (overvenus) [inline src/raftstore/store/fsm/apply.rs]: Could you change it to `Ordering::Acquire`?

File: components/raftstore/src/store/fsm/apply.rs [hunk 8]
```diff
@@ -2658,6 +2702,53 @@ impl ApplyFsm {
...
+                    apply_ctx.engines.kv.clone(),
+                    self.delegate.region.clone(),
+                )),
+            },
+            Err(e) => ReadResponse {
+                response: cmd_resp::new_error(e),
+                snapshot: None,
+            },
+        };
+        cb.invoke_read(resp);
```
Comment (overvenus) [inline components/raftstore/src/store/fsm/apply.rs]: Should we flush write batch before invoking read?

Comment (5kbpers) [reply components/raftstore/src/store/fsm/apply.rs]: Actually we should flush write batch before getting snapshot.

File: components/raftstore/src/coprocessor/mod.rs [hunk 9]
```diff
@@ -159,3 +163,60 @@ pub trait RegionChangeObserver: Coprocessor {
...
+    pub fn is_empty(&self) -> bool {
+        self.cmds.is_empty()
+    }
+}
+
+pub trait CmdObserver: Coprocessor {
+    /// Hook to call after applying write request.
+    fn on_prepare_for_apply(&self, region_id: u64);
+    fn on_apply_cmd(&self, region_id: u64, cmd: Cmd);
+    fn on_flush_apply(&self);
```
Comment (overvenus) [inline components/raftstore/src/coprocessor/mod.rs]: Could you add some documents about them?

File: components/cdc/src/observer.rs [hunk 10]
```diff
@@ -55,6 +59,29 @@ impl CdcObserver {
...
+        self.cmd_batches
+            .borrow_mut()
+            .last_mut()
+            .expect("should exist some cmd batch")
+            .push(region_id, cmd);
+    }
+
+    fn on_flush_apply(&self) {
+        if !self.cmd_batches.borrow().is_empty() {
+            let batches = mem::replace(&mut *self.cmd_batches.borrow_mut(), Vec::default());
```
Comment (overvenus) [inline components/cdc/src/observer.rs]: How about [`RefCell::replace()`](https://doc.rust-lang.org/1.40.0/core/cell/struct.RefCell.html#method.replace)?

**Pull request merged**